File: windows/src/main/token-tracker.ts
/**
 * TokenTracker.swift 포팅
 *
 * %USERPROFILE%\.claude\projects\<encoded-cwd>\<sessionId>.jsonl 을 파싱해
 * 세션별 / 오늘 전체 토큰 사용량을 계산.
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

export interface TokenUsage {
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
}

export const EMPTY_USAGE: TokenUsage = {
  inputTokens: 0,
  outputTokens: 0,
  cacheCreationTokens: 0,
  cacheReadTokens: 0,
};

export function totalTokens(u: TokenUsage): number {
  return u.inputTokens + u.outputTokens + u.cacheCreationTokens + u.cacheReadTokens;
}

export function formatTokens(count: number): string {
  if (count < 1000) return String(count);
  if (count < 1_000_000) return (count / 1000).toFixed(1) + 'K';
  return (count / 1_000_000).toFixed(1) + 'M';
}

export class TokenTracker {
  private projectsDir: string;

  constructor() {
    this.projectsDir = path.join(os.homedir(), '.claude', 'projects');
  }

  /** 특정 세션의 토큰 사용량 */
  usageForSession(sessionId: string): TokenUsage {
    if (!fs.existsSync(this.projectsDir)) return { ...EMPTY_USAGE };

    let projectDirs: string[];
    try {
      projectDirs = fs.readdirSync(this.projectsDir);
    } catch {
      return { ...EMPTY_USAGE };
    }

    for (const dir of projectDirs) {
      const jsonlPath = path.join(this.projectsDir, dir, `${sessionId}.jsonl`);
      if (fs.existsSync(jsonlPath)) {
        return this.parseJSONL(jsonlPath);
      }
    }
    return { ...EMPTY_USAGE };
  }

  /** 오늘 전체 토큰 사용량 (30초 캐싱) */
  private todayCache: TokenUsage | null = null;
  private todayCacheTime = 0;

  todayUsage(): TokenUsage {
    const now = Date.now();
    if (this.todayCache && now - this.todayCacheTime < 30000) {
      return this.todayCache;
    }

    if (!fs.existsSync(this.projectsDir)) return { ...EMPTY_USAGE };

    const total: TokenUsage = { ...EMPTY_USAGE };
    const startOfToday = new Date();
    startOfToday.setHours(0, 0, 0, 0);
    const startMs = startOfToday.getTime();

    let projectDirs: string[];
    try {
      projectDirs = fs.readdirSync(this.projectsDir);
    } catch {
      return total;
    }

    for (const dir of projectDirs) {
      const projectPath = path.join(this.projectsDir, dir);
      let files: string[];
      try {
        files = fs.readdirSync(projectPath);
      } catch {
        continue;
      }

      for (const file of files) {
        if (!file.endsWith('.jsonl')) continue;
        const filePath = path.join(projectPath, file);
        try {
          const stat = fs.statSync(filePath);
          if (stat.mtimeMs < startMs) continue;            // 1차 필터(성능): 오늘 수정된 파일만
          const usage = this.parseJSONL(filePath, startMs); // 2차 필터(정확성): 오늘 줄만
          total.inputTokens += usage.inputTokens;
          total.outputTokens += usage.outputTokens;
          total.cacheCreationTokens += usage.cacheCreationTokens;
          total.cacheReadTokens += usage.cacheReadTokens;
        } catch {
          continue;
        }
      }
    }

    this.todayCache = total;
    this.todayCacheTime = Date.now();
    return total;
  }

  /**
   * JSONL 파싱. sinceMs 를 주면 해당 시각 이후 timestamp 를 가진 줄만 합산한다.
   * (세션 파일 하나가 여러 날에 걸쳐 누적되므로, "오늘" 집계 시 줄 단위로 걸러야 정확하다.)
   */
  private parseJSONL(filePath: string, sinceMs?: number): TokenUsage {
    const usage: TokenUsage = { ...EMPTY_USAGE };
    let content: string;
    try {
      content = fs.readFileSync(filePath, 'utf8');
    } catch {
      return usage;
    }

    for (const line of content.split('\n')) {
      if (!line.trim()) continue;
      try {
        const json = JSON.parse(line);
        const u = json?.message?.usage;
        if (!u) continue;

        // sinceMs 필터: 줄별 timestamp 가 기준 시각 이후인 줄만 (UTC↔로컬은 절대시각 비교라 안전)
        if (sinceMs !== undefined) {
          const t = json?.timestamp ? new Date(json.timestamp).getTime() : NaN;
          if (isNaN(t) || t < sinceMs) continue;
        }

        usage.inputTokens += u.input_tokens || 0;
        usage.outputTokens += u.output_tokens || 0;
        usage.cacheCreationTokens += u.cache_creation_input_tokens || 0;
        usage.cacheReadTokens += u.cache_read_input_tokens || 0;
      } catch {
        continue;
      }
    }

    return usage;
  }

  /** 세션의 모델명 감지 (JSONL에서 마지막 model 필드) */
  modelForSession(sessionId: string): string | null {
    if (!fs.existsSync(this.projectsDir)) return null;

    let projectDirs: string[];
    try {
      projectDirs = fs.readdirSync(this.projectsDir);
    } catch {
      return null;
    }

    for (const dir of projectDirs) {
      const jsonlPath = path.join(this.projectsDir, dir, `${sessionId}.jsonl`);
      if (!fs.existsSync(jsonlPath)) continue;

      let content: string;
      try {
        content = fs.readFileSync(jsonlPath, 'utf8');
      } catch {
        continue;
      }

      // 마지막부터 역순으로 model 필드 찾기
      const lines = content.split('\n').reverse();
      for (const line of lines) {
        if (!line.includes('"model"')) continue;
        try {
          const json = JSON.parse(line);
          const model = json?.message?.model;
          if (typeof model === 'string') return model;
        } catch {
          continue;
        }
      }
    }
    return null;
  }
}

File: Sources/ClaudePet/TokenTracker.swift
import Foundation

struct TokenUsage {
    var inputTokens: Int = 0
    var outputTokens: Int = 0
    var cacheCreationTokens: Int = 0
    var cacheReadTokens: Int = 0

    var totalTokens: Int { inputTokens + outputTokens + cacheCreationTokens + cacheReadTokens }

    static func formatTokens(_ count: Int) -> String {
        if count < 1000 {
            return "\(count)"
        } else if count < 1_000_000 {
            let k = Double(count) / 1000.0
            return String(format: "%.1fK", k)
        } else {
            let m = Double(count) / 1_000_000.0
            return String(format: "%.1fM", m)
        }
    }
}

class TokenTracker {
    private let claudeProjectsDir: String

    init() {
        claudeProjectsDir = NSHomeDirectory() + "/.claude/projects"
    }

    /// 특정 세션의 토큰 사용량
    func usageForSession(_ sessionId: String) -> TokenUsage {
        // 세션 ID로 JSONL 파일 찾기
        let fm = FileManager.default
        guard let projectDirs = try? fm.contentsOfDirectory(atPath: claudeProjectsDir) else {
            return TokenUsage()
        }

        for dir in projectDirs {
            let jsonlPath = "\(claudeProjectsDir)/\(dir)/\(sessionId).jsonl"
            if fm.fileExists(atPath: jsonlPath) {
                return parseJSONL(at: jsonlPath)
            }
        }

        return TokenUsage()
    }

    /// 세션의 모델명 감지
    func modelForSession(_ sessionId: String) -> String? {
        let fm = FileManager.default
        guard let projectDirs = try? fm.contentsOfDirectory(atPath: claudeProjectsDir) else { return nil }

        for dir in projectDirs {
            let jsonlPath = "\(claudeProjectsDir)/\(dir)/\(sessionId).jsonl"
            if fm.fileExists(atPath: jsonlPath),
               let data = fm.contents(atPath: jsonlPath),
               let content = String(data: data, encoding: .utf8) {
                // 마지막부터 역순으로 model 필드 찾기
                for line in content.components(separatedBy: "\n").reversed() {
                    if line.contains("\"model\""),
                       let lineData = line.data(using: .utf8),
                       let json = try? JSONSerialization.jsonObject(with: lineData) as? [String: Any],
                       let message = json["message"] as? [String: Any],
                       let model = message["model"] as? String {
                        return model
                    }
                }
            }
        }
        return nil
    }

    /// 오늘 전체 토큰 사용량
    func todayUsage() -> TokenUsage {
        let fm = FileManager.default
        var total = TokenUsage()

        guard let projectDirs = try? fm.contentsOfDirectory(atPath: claudeProjectsDir) else {
            return total
        }

        let today = Calendar.current.startOfDay(for: Date())

        for dir in projectDirs {
            let projectPath = "\(claudeProjectsDir)/\(dir)"
            guard let files = try? fm.contentsOfDirectory(atPath: projectPath) else { continue }

            for file in files {
                guard file.hasSuffix(".jsonl") else { continue }
                let filePath = "\(projectPath)/\(file)"

                // 1차 필터(성능): 오늘 수정된 파일만 — 그 외 파일엔 오늘 기록이 있을 수 없음
                if let attrs = try? fm.attributesOfItem(atPath: filePath),
                   let modDate = attrs[.modificationDate] as? Date,
                   modDate >= today {
                    // 2차 필터(정확성): 세션 파일은 여러 날에 걸쳐 누적되므로 줄별 timestamp로 오늘 것만 합산
                    let usage = parseJSONL(at: filePath, since: today)
                    total.inputTokens += usage.inputTokens
                    total.outputTokens += usage.outputTokens
                    total.cacheCreationTokens += usage.cacheCreationTokens
                    total.cacheReadTokens += usage.cacheReadTokens
                }
            }
        }

        return total
    }

    // MARK: - JSONL Parsing

    /// JSONL 파싱. `since`를 주면 해당 시각 이후 timestamp를 가진 줄만 합산한다.
    /// (세션 파일 하나가 여러 날에 걸쳐 누적되므로, "오늘" 집계 시 줄 단위로 걸러야 정확하다.)
    private func parseJSONL(at path: String, since: Date? = nil) -> TokenUsage {
        var usage = TokenUsage()

        guard let data = FileManager.default.contents(atPath: path),
              let content = String(data: data, encoding: .utf8) else {
            return usage
        }

        for line in content.components(separatedBy: "\n") {
            guard !line.isEmpty,
                  let lineData = line.data(using: .utf8),
                  let json = try? JSONSerialization.jsonObject(with: lineData) as? [String: Any] else {
                continue
            }

            // message.usage 찾기
            guard let message = json["message"] as? [String: Any],
                  let usageData = message["usage"] as? [String: Any] else {
                continue
            }

            // since 필터: 줄별 timestamp가 기준 시각 이후인 줄만 (UTC ↔ 로컬은 절대시각 비교라 안전)
            if let since = since {
                guard let ts = json["timestamp"] as? String,
                      let date = TokenTracker.parseTimestamp(ts),
                      date >= since else {
                    continue
                }
            }

            usage.inputTokens += usageData["input_tokens"] as? Int ?? 0
            usage.outputTokens += usageData["output_tokens"] as? Int ?? 0
            usage.cacheCreationTokens += usageData["cache_creation_input_tokens"] as? Int ?? 0
            usage.cacheReadTokens += usageData["cache_read_input_tokens"] as? Int ?? 0
        }

        return usage
    }

    // MARK: - Timestamp

    /// ISO8601 timestamp 파싱 — 소수점 초가 있는 형식("…:17.206Z")과 없는 형식 모두 지원
    private static let iso8601WithFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private static let iso8601Plain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()
    private static func parseTimestamp(_ s: String) -> Date? {
        return iso8601WithFractional.date(from: s) ?? iso8601Plain.date(from: s)
    }
}
